storage/objects.go

```diff
@@ -375,6 +485,211 @@ func createDirectory(c ObjectsClient, ctx context.Context, absPath _AbsCleanPath
+	reqInput := client.RequestInput{
+		Method:  http.MethodPost,
+		Path:    "/" + c.client.AccountName + "/uploads",
+		Headers: &http.Header{},
+		Body:    input.Body,
+	}
+	respBody, _, err := c.client.ExecuteRequestStorage(ctx, reqInput)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create mpu")
+	}
+	if respBody != nil {
+		defer respBody.Close()
+	}
+
+	response := &CreateMpuOutput{}
+	decoder := json.NewDecoder(respBody)
+	if err = decoder.Decode(&response); err != nil {
+		return nil, errors.Wrap(err, "unable to decode create mpu response")
+	}
+
+	return response, nil
+}
+
+func getMpu(c ObjectsClient, ctx context.Context, input *GetMpuInput) (*GetMpuOutput, error) {
+	headers := &http.Header{}
+
+	reqInput := client.RequestInput{
+		Method:  http.MethodGet,
+		Path:    input.PartsDirectoryPath + "/state",
+		Headers: headers,
+	}
+	respBody, _, err := c.client.ExecuteRequestStorage(ctx, reqInput)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get mpu")
+	}
+
+	response := &GetMpuOutput{}
+	decoder := json.NewDecoder(respBody)
+	if err = decoder.Decode(&response); err != nil {
+		return nil, errors.Wrap(err, "unable to decode get mpu response")
+	}
+
+	return response, nil
+}
+
+func listMpuParts(c ObjectsClient, ctx context.Context, input *ListMpuPartsInput) (*ListMpuPartsOutput, error) {
+	id := input.Id
+	idLength := len(id)
+	prefixLen, err := strconv.Atoi(id[idLength-1 : idLength])
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to upload part")
+	}
+	prefix := id[:prefixLen]
+	partPath := "/" + c.client.AccountName + "/uploads/" + prefix + "/" + input.Id + "/"
+	listDirInput := ListDirectoryInput{
+		DirectoryName: partPath,
+	}
+
+	dirClient := &DirectoryClient{
+		client: c.client,
+	}
+
+	listDirOutput, err := dirClient.List(ctx, &listDirInput)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to list mpu parts")
+	}
+
+	var parts []ListMpuPart
+	for num, part := range listDirOutput.Entries {
+		parts = append(parts, ListMpuPart{
+			ETag:       part.ETag,
+			PartNumber: num,
+			Size:       int64(part.Size),
+		})
+	}
+
+	listMpuPartsOutput := &ListMpuPartsOutput{
+		Parts: parts,
+	}
+
+	return listMpuPartsOutput, nil
+}
+
+func uploadPart(c ObjectsClient, ctx context.Context, input *UploadPartInput) (*UploadPartOutput, error) {
+	headers := &http.Header{}
+	for key, value := range input.Headers {
+		headers.Set(key, value)
+	}
+
+	if input.ContentMD5 != "" {
+		headers.Set("Content-MD5", input.ContentMD5)
+	}
+	id := input.Id
+	idLength := len(id)
+	partNum := strconv.FormatUint(input.PartNum, 10)
+	prefixLen, err := strconv.Atoi(id[idLength-1 : idLength])
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to upload part")
+	}
+	prefix := id[:prefixLen]
+	partPath := "/" + c.client.AccountName + "/uploads/" + prefix + "/" + input.Id + "/" + partNum
+
+	reqInput := client.RequestNoEncodeInput{
+		Method:  http.MethodPut,
+		Path:    partPath,
+		Headers: &http.Header{},
```

Comment:
Another case of `Headers: headers` here.

Reply:
👍